### PR TITLE
Replace ESNIKeys with ESNIConfig, and delegate the publishing container to HTTPSVC.

### PR DESCRIPTION
**Pull request by chris-wood**
DNSOP adopted HTTPSSVC recently [1]. This is a generalized Resource Record that supports the same functionality as the custom ESNI RRType, including direct IP address specification and ENSI key delivery. This change drops the custom DNS bits from the ESNI document and delegates this to HTTPSSVC, where it probably belongs. This change also renames ESNIKeys to ESNIConfig to more accurately capture its intended purpose, i.e., as a struct for conveying ESNI keys and metadata.

This should address #180, #179, #139, and #110.

[1] https://mailarchive.ietf.org/arch/msg/dnsop/9zCxhCfIhDzA3Cv3D2k4NF_nj4s

### CODE REVIEW

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -158,13 +158,14 @@ innocuous domains, but is also providing protection for some private
...
 
-First, the provider publishes a public key and metadata which is used for SNI
-encryption for all the domains for which it serves directly or indirectly (via
-Split Mode). This document defines a publication mechanism using DNS, but other
-mechanisms are also possible. In particular, if some of the clients of a private
+SNI encryption requires that each provider publish a public key and metadata which
+is used for SNI encryption for all the domains for which it serves directly or
+indirectly (via Split Mode). This document defines the format of the SNI encryption
+public key and metadata, referred to as an ESNI configuration, and delegates DNS
+publication details to {{!I-D.nygren-dnsop-svcb-httpssvc}}, though other delivery
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md]:
```suggestion
publication details to {{!HTTPSSVC=I-D.nygren-dnsop-svcb-httpssvc}}, though other delivery
```

**Comment by ekr** [comment]
I'm not sure that this is going to work with HTTPSSVC as currently defined. The current ESNIKeys record specifies the IP address explicitly and you are supposed to favor that. However, assuming I am reading HTTPSSVC S 6.4 correctly, ipv[46]hint are not in fact favored but are actually disfavored.

  "The "ipv4hint" and "ipv6hint" keys represent IP address hints for the
   service.  If A and AAAA records for SvcDomainName are locally
   available, the client SHOULD ignore these hints.  Otherwise, clients
   MUST perform A and/or AAAA queries for SvcDomainName as in Section 3,
   and clients SHOULD switch to an IP address in those records as soon
   as possible."

This is counter to what we agreed upon for ESNI and resurrects the A/AAAA vs. ESNIKeys mismatch problem. 

Unless I am misunderstanding this, we should first fix HTTPSSVC before making this change.

**Comment by davidben** [comment]
I think the intent with HTTPSSVC is that, if you're using the HTTPSSVC record, you'd look up A/AAAA from the SvcDomainName not the origin name you queried. That way there's still a binding between A/AAAA and ESNIKeys.

This does, however, cost an extra DNS round-trip, so this plan does require the recursive resolvers follow HTTPSSVC for you like they do CNAME, otherwise it would not meet the performance requirements. DoT/DoH are an ecosystem reset, so maybe that's plausible? I think this goes back to how sacred you believe A and AAAA are and whether you believe it's rude to get IP addresses elsewhere. The HTTPSSVC record evidently has different views on this point than the ESNI record.

File: draft-ietf-tls-esni.md [hunk 2]
```diff
@@ -438,7 +294,7 @@ structure:
        ServerESNIResponseType response_type;
        select (response_type) {
            case esni_accept:        uint8 nonce[16];
-           case esni_retry_request: ESNIKeys retry_keys<1..2^16-1>;
+           case esni_retry_request: ESNIConfig retry_keys<1..2^16-1>;
```
Comment (davidben) [inline draft-ietf-tls-esni.md]: Nit: s/retry_keys/retry_configs/

Comment (davidben) [reply draft-ietf-tls-esni.md]: Benefit of the ESNIConfig rename: really `retry_keys` should have been `retry_keyses`, but now we don't have to pretend to be Gollum. :-)

**Comment by chris-wood** [comment]
> This is counter to what we agreed upon for ESNI and resurrects the A/AAAA vs. ESNIKeys mismatch problem.

I would choose to interpret their "SHOULD ignore" advice as optional to ensure the ESNIRecord functionality is maintained. @bemasc, perhaps we can clarify this in the HTTPSVC draft?

**Comment by bemasc** [comment]
> I think the intent with HTTPSSVC is that, if you're using the HTTPSSVC record, you'd look up A/AAAA from the SvcDomainName not the origin name you queried. That way there's still a binding between A/AAAA and ESNIKeys.

Yep, that's exactly right.

> I would choose to interpret their "SHOULD ignore" advice as optional to ensure the ESNIRecord functionality is maintained.

I'm not sure I understand you.  The purpose of "SHOULD ignore" is to ensure that DNS-based load balancing of A/AAAA on SvcDomainName continues to work correctly.  Whether the client uses an ipv6hint or a AAAA record for SvcDomainName, it will be contacting a server that has the corresponding ESNI private key; there's no possibility of mismatch either way.

> perhaps we can clarify this in the HTTPSVC draft?

I think the confusion here is that in HTTPSSVC, clients use the IP of SvcDomainName, not the IP of the origin hostname.  There's no way for a client to accidentally combine the ESNIConfig for the SvcDomainName with the IP of the origin hostname (or of a different SvcDomainName).

I think the draft is pretty clear on this point (e.g. the [intro example](https://tools.ietf.org/html/draft-nygren-dnsop-svcb-httpssvc-00#section-1.1)), but the authors always think that :).  Suggestions are welcome.

**Comment by ekr** [comment]
Well, so as @davidben points out, this requires a second resolution, which is a performance problem.

**Comment by bemasc** [comment]
Yes, waiting for a second resolution is the performance problem that ipv6hint solves, by giving the client an IP address along with the ESNIConfig in the same response.

**Comment by ekr** [comment]
@bemasc: then I just don't see the point of doing the second resolution. It seems disruptive and doesn't add anything. Once I have connected to the server, it can Alt-Svc me itself.

**Comment by ekr** [comment]
I should add, I'm not quite sure how I would implement this in Firefox. Say I resolve example.com and get 1.2.3.4 for ipv4 hints. Now I am supposed to in parallel connect to the server *and* try to resolve SvcDomainName? But what then? I've already started the connection so I'm not going to abort it if I get a 5.6.7.8....

**Comment by bemasc** [comment]
> I just don't see the point of doing the second resolution. It seems disruptive and doesn't add anything.

Consider a CDN customer with their own (old-fashioned, static) DNS.  The customer publishes an HTTPSSVC record whose SvcDomainName, ESNIConfig, and IP hint are the CDN's domain, ESNIConfig, and IP.  Clients connect to the IP from the hint, use this ESNIConfig, and everything works ... but the customer has broken the CDN's load balancing.  All of their traffic is going to a single  CDN IP.

The point of the second resolution is to move customer traffic onto the CDN's preferred IP.

> Once I have connected to the server, it can Alt-Svc me itself.

True, but HTTPSSVC doesn't require clients to support Alt-Svc.  (Chrome, for example, does not support this feature of Alt-Svc.)

> Say I resolve example.com and get 1.2.3.4 for ipv4 hints. Now I am supposed to in parallel connect to the server and try to resolve SvcDomainName? But what then? I've already started the connection so I'm not going to abort it if I get a 5.6.7.8....

The draft says clients "SHOULD switch to an IP address in those records as soon as possible".  An aggressive implementation might start connecting to 5.6.7.8, and close the connection to 1.2.3.4 once it is connected.  (This is roughly what Chrome does for QUIC upgrade.)  A less aggressive implementation might just let the A record sit in the local DNS cache, so that any subsequent connections to this origin can go to the CDN's preferred IP.

**Comment by ekr** [comment]
OK, so I think that this text needs a bunch of work before I would be comfortable with it, as I do not think your "less aggressive implementation" would be conformant with this SHOULD

**Comment by chris-wood** [comment]
@ekr if that’s the case, can you please suggest concrete changes to the text?

**Comment by ekr** [comment]
@bemasc is there some place I can send a PR?

**Comment by bemasc** [comment]
@ekr yes, https://github.com/MikeBishop/dns-alt-svc/pulls

**Comment by ekr** [comment]
https://github.com/MikeBishop/dns-alt-svc/pull/69

**Comment by ekr** [comment]
OK, the new text has been merged, so I am now OK with this PR.

**Pull request merged**